design/kopia.md

```diff
@@ -0,0 +1,461 @@
+
+The `spec.policy` property can be used to change the snapshot settings of the
+respository at the global, user or directory levels. It maps to the
+options supported by the `kopia policy set [target]` [command][2].
+
+The `status` subresource provides conditions and status information to ensure
+that the controller does not attempt to re-create the repository, nor server
+during restart.
+
+## Repository Lifecycle
+
+When a `Repository` resource is created, the `RepositoryController` controller
+responds by creating a new Kopia repository using the equivalent of this
+command:
+
+```sh
+kopia repository create s3 \
+  --bucket=my-bucket \
+  --access-key=<redacted> \
+  --secret-access-key=<redacted> \
+  --create-only=true \
+  [--endpoint=https://foo.example.com \]
+  [--prefix=my-prefix \]
+  [--region=us-west-1]
+```
+
+> 📝 The examples in this section uses S3 for illustration purposes. Other
+> supported storage types include GCS And Azure Blob Storage.
+
+### Server-Side Setup
+
+Once the repository is ready, the controller will schedule a Kopia server
+instance to run in the `kanister` namespace, as a `Deployment` workload. The
+server is started using the equivalent of this command:
+
+```sh
+kopia server start --address=0.0.0.0:51515 \
+  --config-file=/run/kopia/repo.config \
+  --tls-cert-file=/run/kopia/tls.crt \
+  --tls-key-file=/run/kopia/tls.key \
+  --server-username=<redacted> \
+  --server-password=<redacted> \
+  --server-control-username=<redacted> \
+  --server-control-password=<redacted> \
+```
+
+The `/run/kopia/repo.config` configuration file is generated from the data found
+in the secret referenced by the repository's `spec.location.secretName` property.
+See the [Kopia documentation][3] and [GitHub source code][4] for more
+information on the configuration file format, and supported configuration.
+
+The `/run/kopia/tls.crt` and `/run/kopia/tls.key` files contain the TLS x509
+certificate and private key read from the secret referenced by the repository's
+`spec.server.tlsSecretName` property.
+
+The credentials for the `--server-username`, `--server-password`,
+`--server-control-username` and `--server-control-password` options are read
+from the secret referenced by the `spec.server.adminSecretName` property.
+
+### Client-Side Setup
+
+The Kopia server is fronted by a K8s `Service` resource. Data mover clients will
+connect to it using the equivalence of:
+
+```sh
+kopia repository connect server \
+  --url=https://<service-name>.kanister.svc.cluster.local \
+  --config-file=/run/kopia/repo.config \
+  --server-cert-fingerprint=<redacted> \
+  --password=<redacted>
+```
+
+The `<service-name>` is the Kopia server's `Service` resource name. By
+convention, it includes the repository's `targetedNamespace`. E.g., the
+`Service` fronting the Kopia server of a repository serving the
+`monitoring` namespace can be `repo-monitoring`. Then its FQDN is
+`repo-monitoring.kanister.svc.cluster.local`.
```

Comment:
What happens when the user wants to connect to a different repository on the same back-end location and operate on same application (`targetNamespace`)? 

If a new `Repository` CR is expected to be created and used here, how would the new CR differentiate between `Service` resources that include same `targetNamespace`? I think the resultant FQDN would be a point of conflict between such multiple `Repository` CRs.

Reply:
Can they use a different `<service-name>` e.g., `repo2-monitoring`?

Reply:
Yes. The `Service` resources can have some prefix or suffix to differentiate. Probably, we can use generate name in the `repository-controller` with the `<service-name>` to do that. But then, we would need to devise a strategy for passing the resultant `<service-name>` (or entire FQDN) to the data mover clients.